### PR TITLE
fix(clients/java): make a `ZeebeCallCredentials#applyRequestMetadata`call non-blocking

### LINKED ISSUE #11816: `ZeebeCallCredentials#applyRequestMetadata` should not block a thread
**Describe the bug**

[The Javadoc](https://github.com/grpc/grpc-java/blob/26c2db9c39f71c6b507c8066c5f41a23ed7af986/api/src/main/java/io/grpc/CallCredentials.java#L43-L47) says:
> <p>It is called for each individual RPC, within the {@link Context} of the call, before the
>  stream is about to be created on a transport. Implementations should not block in this
>  method. If metadata is not immediately available, e.g., needs to be fetched from network, the
>  implementation may give the {@code applier} to an asynchronous task which will eventually call
>  the {@code applier}. The RPC proceeds only after the {@code applier} is called.

But current implementation is blocking ([here](https://github.com/camunda/zeebe/blob/571442be0484468859df5e47ed1b50efea5db744/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeCallCredentials.java#L46) and [here](https://github.com/camunda/zeebe/blob/main/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java#L176-L190)).

**To Reproduce**

I can not see a clear way to reproduce it, but if you apply this patch:
```patch
Index: clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java b/clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java
--- a/clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java	(revision 571442be0484468859df5e47ed1b50efea5db744)
+++ b/clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java	(date 1677255549375)
@@ -21,6 +21,7 @@
 import static org.mockito.Mockito.times;
 
 import io.camunda.zeebe.client.api.command.ClientException;
+import io.camunda.zeebe.client.impl.Loggers;
 import io.camunda.zeebe.client.impl.ZeebeClientBuilderImpl;
 import io.camunda.zeebe.client.impl.ZeebeClientImpl;
 import io.camunda.zeebe.client.util.RecordingGatewayService;
@@ -68,12 +69,14 @@
     final String bearerToken = "Bearer someToken";
     final ZeebeClientBuilderImpl builder = new ZeebeClientBuilderImpl();
 
+    Loggers.LOGGER.info("Current thread: {}", Thread.currentThread().getName());
     builder
         .usePlaintext()
         .credentialsProvider(
             new CredentialsProvider() {
               @Override
               public void applyCredentials(final Metadata headers) {
+                Loggers.LOGGER.info("Current thread: {}", Thread.currentThread().getName());
                 headers.put(AUTH_KEY, bearerToken);
               }
 

```

And run this test, you would see that thread names are the same:
```
21:25:32.199 [main] INFO  io.camunda.zeebe.client - Current thread: main
21:25:32.241 [main] INFO  io.camunda.zeebe.client - Current thread: main
```

**Expected behavior**

I expect that the rule from the javadoc will be applied and the `CallCredentials` thread are gain from the GRPC thread pool.

**Environment:**
- OS: MacOS 
- Zeebe Version: 8.1.6 (all known versions has this bug) 
- Configuration: Just `ZeebeClient`

### PR DESCRIPTION
**Pull request by aivinog1**
## Description

Based on the Javadoc, this method shouldn't be blocking.

## Related issues

closes #11816 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [X] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by aivinog1** [comment]
Hey reviewer!
Can you please backport it in 8.1 and 8.0? :)

Thank you :)

**Comment by remcowesterhoud** [comment]
Hey @aivinog1, thanks for the PR! As you can see the CI is broken 😄 It should be fixed on main currently. Could you please do a rebase so I can trigger it again and see if the fix worked 🤞

**Comment by aivinog1** [comment]
Hey @remcowesterhoud! Thanks, it's done.

File: clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java [hunk 1]
```diff
@@ -175,4 +177,36 @@ public void shouldNotChangeHeadersWithNoProvider() {
     // then
     assertThat(recordingInterceptor.getCapturedHeaders().containsKey(AUTH_KEY)).isFalse();
   }
+
+  @Test
+  public void shouldCredentialsProviderRunFromGRPCThreadPool() {
```
Comment (remcowesterhoud) [inline clients/java/src/test/java/io/camunda/zeebe/client/CredentialsTest.java:182]: 👍 Clever how you wrote this test case. I like it!

**Review by remcowesterhoud — APPROVED** [review]
Thanks for the fix @aivinog1. I'm curious how you noticed this. Did this impact you in any way?

bors merge

**Comment by aivinog1** [comment]
> Thanks for the fix @aivinog1. I'm curious how you noticed this. Did this impact you in any way?

Well, I'm porting workers from Zeebe Client Java to [the Kotlin + Coroutines stack](https://github.com/camunda-community-hub/kotlin-coworker). And I found out that there was a blocking operation inside the calls. So, I decided to double-check it here and found out the problem :) 

P.S. If you are curious why I'm porting workers, you can check [the plain and simple performance comparison test](https://github.com/camunda-community-hub/kotlin-coworker/blob/1ce1e3e9b7ce2a203c4fd41ddc6c698e2958b370/benchmark/src/test/kotlin/org/camunda/community/extension/coworker/CoworkerVsJobWorkerPerformanceComparisonTest.kt#L41-L137). Same worker but in the reactive stack can process 40 jobs 20+ times faster :) It is still in the PoC phase but solves the needs of scaling workers for performance reasons.

In my machine the numbers are next:
```
For Zeebe Client it took 41.186149961s to process 40
...
For Coworker it took 1.522231647s to process 40
So, Zeebe Client Java duration / Coworker duration = 27.056427346106805
```

P.P.S. My project is still pretty ugly inside 🙈

**Comment by remcowesterhoud** [comment]
Very cool, thanks for sharing!

**Pull request merged**